File: flutter_app/lib/services/agent_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:supabase_flutter/supabase_flutter.dart';
import '../core/config.dart';
import '../models/agent_model.dart';

class AgentService {
  final SupabaseClient _supabase = Supabase.instance.client;

  String? get _userId => _supabase.auth.currentUser?.id;

  // ─── CRUD ───

  Future<List<Agent>> getAgents() async {
    if (_userId == null) return [];
    final data = await _supabase
        .from('agents')
        .select()
        .eq('user_id', _userId!)
        .order('created_at', ascending: false);
    return (data as List).map((e) => Agent.fromJson(e)).toList();
  }

  Future<Agent?> createAgent(CreateAgentInput input) async {
    if (_userId == null) return null;
    final data = await _supabase
        .from('agents')
        .insert({...input.toJson(), 'user_id': _userId})
        .select()
        .single();
    return Agent.fromJson(data);
  }

  Future<bool> updateAgent(String id, Map<String, dynamic> updates) async {
    try {
      await _supabase.from('agents').update(updates).eq('id', id);
      return true;
    } catch (e) {
      debugPrint('Update agent error: $e');
      return false;
    }
  }

  Future<bool> deleteAgent(String id) async {
    try {
      // Delete conversations and messages first
      final convs = await _supabase
          .from('agent_conversations')
          .select('id')
          .eq('agent_id', id);
      for (final conv in convs) {
        await _supabase
            .from('agent_messages')
            .delete()
            .eq('conversation_id', conv['id']);
      }
      await _supabase.from('agent_conversations').delete().eq('agent_id', id);
      await _supabase.from('agents').delete().eq('id', id);
      return true;
    } catch (e) {
      debugPrint('Delete agent error: $e');
      return false;
    }
  }

  // ─── Conversations ───

  Future<String?> getOrCreateConversation(String agentId) async {
    if (_userId == null) return null;

    final convs = await _supabase
        .from('agent_conversations')
        .select('id')
        .eq('agent_id', agentId)
        .eq('user_id', _userId!)
        .order('updated_at', ascending: false)
        .limit(1);

    if (convs.isNotEmpty) return convs[0]['id'] as String;

    final data = await _supabase
        .from('agent_conversations')
        .insert({
          'agent_id': agentId,
          'user_id': _userId,
          'title': 'Chat',
        })
        .select()
        .single();
    return data['id'] as String;
  }

  Future<List<AgentMessage>> getMessages(String conversationId) async {
    final data = await _supabase
        .from('agent_messages')
        .select()
        .eq('conversation_id', conversationId)
        .order('created_at', ascending: true);
    return (data as List).map((e) => AgentMessage.fromJson(e)).toList();
  }

  Future<void> saveMessage({
    required String conversationId,
    required String role,
    required String content,
  }) async {
    await _supabase.from('agent_messages').insert({
      'conversation_id': conversationId,
      'role': role,
      'content': content,
    });
  }

  // ─── Chat streaming ───

  Future<void> streamChat({
    required String agentId,
    required List<Map<String, String>> messages,
    required void Function(String chunk) onChunk,
    required void Function() onDone,
    required void Function(String error) onError,
  }) async {
    try {
      final session = _supabase.auth.currentSession;
      if (session == null) {
        onError('Not authenticated');
        return;
      }

      final url = Uri.parse(
        '${AppConfig.supabaseUrl}/functions/v1/agent-chat',
      );

      final request = http.Request('POST', url);
      request.headers.addAll({
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${session.accessToken}',
        'apikey': AppConfig.supabaseAnonKey,
      });
      request.body = jsonEncode({
        'agentId': agentId,
        'messages': messages,
      });

      final client = http.Client();
      final response = await client.send(request);

      if (response.statusCode != 200) {
        final body = await response.stream.bytesToString();
        try {
          final err = jsonDecode(body);
          onError(err['error'] ?? 'Chat failed');
        } catch (_) {
          onError('Chat failed (${response.statusCode})');
        }
        client.close();
        return;
      }

      final completer = Completer<void>();
      String buffer = '';

      response.stream
          .transform(utf8.decoder)
          .listen(
            (chunk) {
              buffer += chunk;
              int newlineIdx;
              while ((newlineIdx = buffer.indexOf('\n')) != -1) {
                String line = buffer.substring(0, newlineIdx);
                buffer = buffer.substring(newlineIdx + 1);
                if (line.endsWith('\r')) line = line.substring(0, line.length - 1);
                if (line.startsWith(':') || line.trim().isEmpty) continue;
                if (!line.startsWith('data: ')) continue;
                final jsonStr = line.substring(6).trim();
                if (jsonStr == '[DONE]') continue;
                try {
                  final parsed = jsonDecode(jsonStr);
                  final content = parsed['choices']?[0]?['delta']?['content'];
                  if (content != null && content is String) {
                    onChunk(content);
                  }
                } catch (_) {}
              }
            },
            onDone: () {
              onDone();
              client.close();
              completer.complete();
            },
            onError: (e) {
              onError(e.toString());
              client.close();
              completer.complete();
            },
          );

      await completer.future;
    } catch (e) {
      onError(e.toString());
    }
  }

  // ─── Agent statuses ───

  Future<Map<String, String>> fetchStatuses() async {
    try {
      final session = _supabase.auth.currentSession;
      if (session == null) return {};

      final url = Uri.parse(
        '${AppConfig.supabaseUrl}/functions/v1/agent-status',
      );
      final resp = await http.get(url, headers: {
        'Authorization': 'Bearer ${session.accessToken}',
        'apikey': AppConfig.supabaseAnonKey,
      });

      if (resp.statusCode == 200) {
        final data = jsonDecode(resp.body);
        if (data['statuses'] != null) {
          final map = <String, String>{};
          (data['statuses'] as Map<String, dynamic>).forEach((id, info) {
            map[id] = (info as Map<String, dynamic>)['status'] ?? 'unknown';
          });
          return map;
        }
      }
    } catch (e) {
      debugPrint('Fetch statuses error: $e');
    }
    return {};
  }

  // ─── RunPod ───

  Future<String?> createRunpodEndpoint(String agentId, String template) async {
    try {
      final resp = await _supabase.functions.invoke(
        'create-runpod-endpoint',
        body: {'agentId': agentId, 'template': template},
      );
      return resp.data?['endpointId'] as String?;
    } catch (e) {
      debugPrint('RunPod provision error: $e');
      return null;
    }
  }

  Future<void> deleteRunpodEndpoint(String agentId) async {
    try {
      await _supabase.functions.invoke(
        'delete-runpod-endpoint',
        body: {'agentId': agentId},
      );
    } catch (e) {
      debugPrint('RunPod cleanup error: $e');
    }
  }
}

File: src/services/agentService.ts
import { supabase, TIMELESS_SUPABASE_URL, TIMELESS_ANON_KEY } from "@/lib/supabase";

export interface Agent {
  id: string;
  name: string;
  role: string | null;
  system_prompt: string | null;
  tools: string[];
  model: string | null;
  avatar_url: string | null;
  paused: boolean;
  runpod_endpoint_id: string | null;
  runpod_model: string | null;
  user_id: string;
  created_at: string;
  updated_at: string;
}

export interface CreateAgentInput {
  name: string;
  role?: string;
  system_prompt?: string;
  tools?: string[];
  model?: string;
}

export interface AgentMessage {
  id: string;
  role: string;
  content: string;
  created_at?: string;
}

// ─── CRUD ───

export async function getAgents(): Promise<Agent[]> {
  const user = supabase.auth.getUser ? (await supabase.auth.getUser()).data.user : null;
  if (!user) return [];

  const { data, error } = await supabase
    .from("agents")
    .select("*")
    .eq("user_id", user.id)
    .order("created_at", { ascending: false });

  if (error) {
    console.error("getAgents error:", error);
    return [];
  }
  return (data ?? []) as Agent[];
}

export async function createAgent(input: CreateAgentInput): Promise<Agent | null> {
  const user = (await supabase.auth.getUser()).data.user;
  if (!user) return null;

  const { data, error } = await supabase
    .from("agents")
    .insert({
      ...input,
      tools: input.tools ?? [],
      user_id: user.id,
    })
    .select()
    .single();

  if (error) {
    console.error("createAgent error:", error);
    return null;
  }
  return data as Agent;
}

export async function updateAgent(id: string, updates: Partial<Agent>): Promise<boolean> {
  const { error } = await supabase.from("agents").update(updates).eq("id", id);
  return !error;
}

export async function deleteAgent(id: string): Promise<boolean> {
  try {
    // Delete conversations and messages first
    const { data: convs } = await supabase
      .from("agent_conversations")
      .select("id")
      .eq("agent_id", id);

    if (convs) {
      for (const conv of convs) {
        await supabase.from("agent_messages").delete().eq("conversation_id", conv.id);
      }
    }
    await supabase.from("agent_conversations").delete().eq("agent_id", id);
    await supabase.from("agents").delete().eq("id", id);
    return true;
  } catch (e) {
    console.error("deleteAgent error:", e);
    return false;
  }
}

// ─── Conversations ───

export async function getOrCreateConversation(agentId: string): Promise<string | null> {
  const user = (await supabase.auth.getUser()).data.user;
  if (!user) return null;

  const { data: convs } = await supabase
    .from("agent_conversations")
    .select("id")
    .eq("agent_id", agentId)
    .eq("user_id", user.id)
    .order("updated_at", { ascending: false })
    .limit(1);

  if (convs && convs.length > 0) return convs[0].id;

  const { data, error } = await supabase
    .from("agent_conversations")
    .insert({ agent_id: agentId, user_id: user.id, title: "Chat" })
    .select()
    .single();

  if (error) return null;
  return data.id;
}

export async function getMessages(conversationId: string): Promise<AgentMessage[]> {
  const { data } = await supabase
    .from("agent_messages")
    .select("*")
    .eq("conversation_id", conversationId)
    .order("created_at", { ascending: true });

  return (data ?? []) as AgentMessage[];
}

export async function saveMessage(conversationId: string, role: string, content: string) {
  await supabase.from("agent_messages").insert({
    conversation_id: conversationId,
    role,
    content,
  });
}

// ─── Chat streaming via SSE ───

export async function streamChat(
  agentId: string,
  messages: { role: string; content: string }[],
  onChunk: (chunk: string) => void,
  onDone: () => void,
  onError: (error: string) => void
) {
  const session = (await supabase.auth.getSession()).data.session;
  if (!session) {
    onError("Not authenticated");
    return;
  }

  try {
    const response = await fetch(`${TIMELESS_SUPABASE_URL}/functions/v1/agent-chat`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${session.access_token}`,
        apikey: TIMELESS_ANON_KEY,
      },
      body: JSON.stringify({ agentId, messages }),
    });

    if (!response.ok) {
      const text = await response.text();
      try {
        const err = JSON.parse(text);
        onError(err.error || "Chat failed");
      } catch {
        onError(`Chat failed (${response.status})`);
      }
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) {
      onError("No response stream");
      return;
    }

    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith(":")) continue;
        if (!trimmed.startsWith("data: ")) continue;
        const jsonStr = trimmed.substring(6).trim();
        if (jsonStr === "[DONE]") continue;
        try {
          const parsed = JSON.parse(jsonStr);
          const content = parsed?.choices?.[0]?.delta?.content;
          if (content) onChunk(content);
        } catch {}
      }
    }

    onDone();
  } catch (e: any) {
    onError(e.message || "Stream error");
  }
}
